cohttp-lwt/src/s.ml

```diff
@@ -106,13 +106,27 @@ module type Server = sig
 
   type conn = IO.conn * Cohttp.Connection.t
 
+  type response_action =
+    (* The connection is not closed in the [`Expert] case until the [unit
+       Lwt.t] becomes determined. *)
+    [ `Expert of Cohttp.Response.t
+                 * (IO.ic
+                    -> IO.oc
+                    -> unit Lwt.t)
```

Comment:
Should we also define what happens if this function leaks an exception?  Where will that be dispatched to?

Reply:
I agree - we should encourage the developer to treat leaking of exceptions as an internal server error, and not part of the published API.  We do need to be careful to generate an information-free 500 (i.e. not put the error in the return body of the 500, since that might leak sensitive information or be vulnerable to CRLF)

Reply:
Thinking about this again, it may not make sense write a 500 response to the client in case of an exception. If the function `IO.ic -> IO.oc -> unit Lwt.t` raises an exception, it's impossible to know what state the client/server are in. As an example, if the client and server have switched to websocket protocol, it would be incorrect to write a 500 HTTP response to the client. Bubbling the exception up to `on_exn` and closing the connection may be the best option.

Am I overlooking something?

Reply:
That's true. I'm fine with just logging and closing the connection. The more conservative the better in this situation :)